### PR TITLE
EOS-26970 : Modify system health logic remove usage of timeout while reporting alerts to hare

### PR DESCRIPTION
**Pull request by akash2144**
# Problem Statement
- Modify system health logic remove usage of timeout while reporting alerts to hare
# Design
-  system health bootstrap currently uses timeout functionality to get the initial health of all the pods(nodes).
This timeout needs to be removed. Instead following logic should be used to send alerts to Hare:
If new consul key needs to be added to system health for the recieved alert from k8s monitor, no alert should be sent to hare. If event from k8s monitor received for existing key, alert should be generated (if applicable) and sent to hare

# Coding
-  [x] Coding conventions are followed and code is consistent

# Testing 
- [x] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [x] Testing was performed with RPM

# Review Checklist 
- [x] PR is self reviewed
- [x] JIRA number/GitHub Issue added to PR
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [x] Check if the description is clear and explained
- [ ] Is there a change in filename/package/module or signature? [Y/N]: N
- [ ] If yes for above point, is a notification sent to all other cortx components? [Y/N] NA
- [ ] Side effects on other features (deployment/upgrade)? [Y/N] N
- [ ] Dependencies on other component(s)? [Y/N] N
-     If yes for above point, post link to the corresponding PR.

# Review Checklist 
- [ ] Is perfline test run and the report with and without the changes updated in the PR? [Y/N]: 

# Documentation
  Checklist for Author
- [x] Changes done to WIKI / Confluence page / Quick Start Guide

# Testing Details
<pre>
  <code>
    <h3># During bootstrap time period, When all 3 services are started. No alerts sent to Hare.
    <h4>[root@cortx-control-headless-svc /]# tail -f /var/log/seagate/cortx/ha/pod_monitor.log
2022-01-06 06:02:57 pod_monitor [7557]: INFO [__init__] Init done for pod monitor
2022-01-06 06:02:57 pod_monitor [7557]: INFO [run] Sending alert on message bus {'_resource_type': 'host', '_resource_name': 'ssc-vm-g2-rhev4-2987.colo.seagate.com', '_event_type': 'online', '_k8s_container': None, '_generation_id': None, '_node': None, '_is_status': True, '_timestamp': '1641448977'}
2022-01-06 06:02:58 pod_monitor [7557]: INFO [run] Sending alert on message bus {'_resource_type': 'host', '_resource_name': 'ssc-vm-g2-rhev4-2988.colo.seagate.com', '_event_type': 'online', '_k8s_container': None, '_generation_id': None, '_node': None, '_is_status': True, '_timestamp': '1641448978'}
2022-01-06 06:02:58 pod_monitor [7557]: INFO [run] Sending alert on message bus {'_resource_type': 'node', '_resource_name': '6cbc441cad6b41e1a5d7bbd5fc3e5903', '_event_type': 'online', '_k8s_container': None, '_generation_id': 'cortx-data-pod-ssc-vm-g2-rhev4-2988-588d9d4fc7-fz65v', '_node': 'ssc-vm-g2-rhev4-2988.colo.seagate.com', '_is_status': True, '_timestamp': '1641448978'}
2022-01-06 06:02:58 pod_monitor [7557]: INFO [run] Sending alert on message bus {'_resource_type': 'host', '_resource_name': 'ssc-vm-g2-rhev4-2989.colo.seagate.com', '_event_type': 'online', '_k8s_container': None, '_generation_id': None, '_node': None, '_is_status': True, '_timestamp': '1641448978'}
2022-01-06 06:02:58 pod_monitor [7557]: INFO [run] Sending alert on message bus {'_resource_type': 'node', '_resource_name': 'ead903f9bf9042e3b27c276291fff38c', '_event_type': 'online', '_k8s_container': None, '_generation_id': 'cortx-data-pod-ssc-vm-g2-rhev4-2987-5cdc6664d4-rfmkl', '_node': 'ssc-vm-g2-rhev4-2987.colo.seagate.com', '_is_status': True, '_timestamp': '1641448978'}
2022-01-06 06:02:58 pod_monitor [7557]: INFO [run] Sending alert on message bus {'_resource_type': 'node', '_resource_name': '27d9b51227854380b0c3459a8f713bd7', '_event_type': 'online', '_k8s_container': None, '_generation_id': 'cortx-data-pod-ssc-vm-g2-rhev4-2989-79b5c57669-nnv4d', '_node': 'ssc-vm-g2-rhev4-2989.colo.seagate.com', '_is_status': True, '_timestamp': '1641448978'}

[root@cortx-control-headless-svc /]# tail -f /var/log/seagate/cortx/ha/health_monitor.log
2022-01-06 06:02:56 health_monitor [7555]: INFO [__init__] Health Monitor daemon initializations...
2022-01-06 06:02:56 health_monitor [7555]: INFO [run] Running the daemon for Health Monitor with PID 7555...
2022-01-06 06:02:56 health_monitor [7555]: INFO [__init__] MessageBus initialized as kafka

    <h3># Thereafter, When one of the cortx-data pod deleted, Event is sent to Hare:
    <h4>[root@ssc-vm-g2-rhev4-2987 ~]# kubectl delete pod cortx-data-pod-ssc-vm-g2-rhev4-2989-79b5c57669-nnv4d

[root@cortx-control-headless-svc /]# tail -f /var/log/seagate/cortx/ha/pod_monitor.log
2022-01-06 06:07:48 pod_monitor [7557]: INFO [run] Sending alert on message bus {'_resource_type': 'node', '_resource_name': '27d9b51227854380b0c3459a8f713bd7', '_event_type': 'online', '_k8s_container': None, '_generation_id': 'cortx-data-pod-ssc-vm-g2-rhev4-2989-79b5c57669-jbxwv', '_node': 'ssc-vm-g2-rhev4-2989.colo.seagate.com', '_is_status': False, '_timestamp': '1641449268'}
2022-01-06 06:08:11 pod_monitor [7557]: INFO [run] Sending alert on message bus {'_resource_type': 'node', '_resource_name': '27d9b51227854380b0c3459a8f713bd7', '_event_type': 'failed', '_k8s_container': None, '_generation_id': 'cortx-data-pod-ssc-vm-g2-rhev4-2989-79b5c57669-nnv4d', '_node': 'ssc-vm-g2-rhev4-2989.colo.seagate.com', '_is_status': False, '_timestamp': '1641449291'}

[root@cortx-control-headless-svc /]# tail -f /var/log/seagate/cortx/ha/health_monitor.log
2022-01-06 06:02:56 health_monitor [7555]: INFO [__init__] Health Monitor daemon initializations...
2022-01-06 06:02:56 health_monitor [7555]: INFO [run] Running the daemon for Health Monitor with PID 7555...
2022-01-06 06:02:56 health_monitor [7555]: INFO [__init__] MessageBus initialized as kafka
2022-01-06 06:08:13 health_monitor [7555]: INFO [evaluate] Evaluated action ['publish'] for key action/node/failed
2022-01-06 06:08:13 health_monitor [7555]: INFO [process_event] Evaluated {"event_id": "1641449292cea6b04bf6c6402890b166d3fd8e11fd", "event_type": "failed", "severity": "error", "site_id": "1", "rack_id": "1", "cluster_id": "eabfcf05cf3f4c0db4fdde7ba7ae1285", "storageset_id": "27d9b51227854380b0c3459a8f713bd7", "node_id": "27d9b51227854380b0c3459a8f713bd7", "host_id": "27d9b51227854380b0c3459a8f713bd7", "resource_type": "node", "timestamp": "1641449291", "resource_id": "27d9b51227854380b0c3459a8f713bd7", "specific_info": "specific_info"} with action ['publish']
2022-01-06 06:08:13 health_monitor [7555]: INFO [act] Default action handler with Event: 1641449292cea6b04bf6c6402890b166d3fd8e11fd and actions : ['publish']
2022-01-06 06:08:13 health_monitor [7555]: INFO [publish] Sending action event {"version": "2.0", "event_type": "failed", "event_id": "1641449292cea6b04bf6c6402890b166d3fd8e11fd", "resource_type": "node", "cluster_id": "eabfcf05cf3f4c0db4fdde7ba7ae1285", "site_id": "1", "rack_id": "1", "storageset_id": "27d9b51227854380b0c3459a8f713bd7", "node_id": "27d9b51227854380b0c3459a8f713bd7", "resource_id": "27d9b51227854380b0c3459a8f713bd7", "timestamp": "1641449291", "event_specific_info": "specific_info"} to component hare
2022-01-06 06:08:14 health_monitor [7555]: INFO [evaluate] Evaluated action [] for key action/rack/degraded
2022-01-06 06:08:14 health_monitor [7555]: INFO [evaluate] Evaluated action [] for key action/site/degraded
2022-01-06 06:08:14 health_monitor [7555]: INFO [evaluate] Evaluated action [] for key action/cluster/degraded

    <h3>Created test-pod & started, During bootstrap of test-pod No events are sent to hare
    <h4>[root@cortx-control-headless-svc /]# tail -f /var/log/seagate/cortx/ha/pod_monitor.log
2022-01-06 11:38:58 pod_monitor [8764]: INFO [run] Sending alert on message bus {'_resource_type': 'node', '_resource_name': '27d9b5122785444444444444444', '_event_type': 'online', '_k8s_container': None, '_generation_id': 'test-pod', '_node': 'ssc-vm-g2-rhev4-2988.colo.seagate.com', '_is_status': False, '_timestamp': '1641469138'}

No new alert seen in health_monitor.log
[root@cortx-control-headless-svc /]# tail -f /var/log/seagate/cortx/ha/health_monitor.log

    <h3># After bootstrap of test-pod, Deleted test-pod. Events are sent to Hare
    <h4>[root@cortx-control-headless-svc /]# tail -f /var/log/seagate/cortx/ha/pod_monitor.log
2022-01-06 11:41:23 pod_monitor [8764]: INFO [run] Sending alert on message bus {'_resource_type': 'node', '_resource_name': '27d9b5122785444444444444444', '_event_type': 'failed', '_k8s_container': None, '_generation_id': 'test-pod', '_node': 'ssc-vm-g2-rhev4-2988.colo.seagate.com', '_is_status': False, '_timestamp': '1641469283'}

[root@cortx-control-headless-svc /]# tail -f /var/log/seagate/cortx/ha/health_monitor.log
2022-01-06 11:41:24 health_monitor [8762]: INFO [evaluate] Evaluated action ['publish'] for key action/node/failed
2022-01-06 11:41:24 health_monitor [8762]: INFO [process_event] Evaluated {"event_id": "164146928395bd0bbd00a44dedb0aeb62750b9f0b0", "event_type": "failed", "severity": "error", "site_id": "1", "rack_id": "1", "cluster_id": "eabfcf05cf3f4c0db4fdde7ba7ae1285", "storageset_id": "27d9b5122785444444444444444", "node_id": "27d9b5122785444444444444444", "host_id": "27d9b5122785444444444444444", "resource_type": "node", "timestamp": "1641469283", "resource_id": "27d9b5122785444444444444444", "specific_info": "specific_info"} with action ['publish']
2022-01-06 11:41:24 health_monitor [8762]: INFO [act] Default action handler with Event: 164146928395bd0bbd00a44dedb0aeb62750b9f0b0 and actions : ['publish']
2022-01-06 11:41:24 health_monitor [8762]: INFO [publish] Sending action event {"version": "2.0", "event_type": "failed", "event_id": "164146928395bd0bbd00a44dedb0aeb62750b9f0b0", "resource_type": "node", "cluster_id": "eabfcf05cf3f4c0db4fdde7ba7ae1285", "site_id": "1", "rack_id": "1", "storageset_id": "27d9b5122785444444444444444", "node_id": "27d9b5122785444444444444444", "resource_id": "27d9b5122785444444444444444", "timestamp": "1641469283", "event_specific_info": "specific_info"} to component hare

    <h3># Also confirmed that "sys_health_bootstrap_timeout" key is not present in ha.conf file. Consul Keys are
    <h4>[root@cortx-control-headless-svc /]# consul kv get --http-addr=consul-server-0.consul-server.default.svc.cluster.local:8500 --recurse cortx/ha
cortx/ha/v1:
cortx/ha/v1/action/node/failed:["publish"]
cortx/ha/v1/action/node/online:["publish"]
cortx/ha/v1/cortx/ha/system/cluster/eabfcf05cf3f4c0db4fdde7ba7ae1285/health:{"events": [{"event_timestamp": "1641468928", "created_timestamp": "1641468935", "status": "degraded", "specific_info": null}, {"event_timestamp": "1641468927", "created_timestamp": "1641468930", "status": "online", "specific_info": null}], "action": {"modified_timestamp": "1641468935", "status": "pending"}, "attributes": {}}
cortx/ha/v1/cortx/ha/system/cluster/eabfcf05cf3f4c0db4fdde7ba7ae1285/site/1/health:{"events": [{"event_timestamp": "1641468928", "created_timestamp": "1641468934", "status": "degraded", "specific_info": null}, {"event_timestamp": "1641468927", "created_timestamp": "1641468929", "status": "online", "specific_info": null}], "action": {"modified_timestamp": "1641468934", "status": "pending"}, "attributes": {}}
cortx/ha/v1/cortx/ha/system/cluster/eabfcf05cf3f4c0db4fdde7ba7ae1285/site/1/rack/1/health:{"events": [{"event_timestamp": "1641468928", "created_timestamp": "1641468933", "status": "degraded", "specific_info": null}, {"event_timestamp": "1641468927", "created_timestamp": "1641468929", "status": "online", "specific_info": null}], "action": {"modified_timestamp": "1641468933", "status": "pending"}, "attributes": {}}
cortx/ha/v1/cortx/ha/system/cluster/eabfcf05cf3f4c0db4fdde7ba7ae1285/site/1/rack/1/node/27d9b51227854380b0c3459a8f713bd7/health:{"events": [{"event_timestamp": "1641469450", "created_timestamp": "1641469450", "status": "online", "specific_info": "specific_info"}, {"event_timestamp": "1641468928", "created_timestamp": "1641468933", "status": "failed", "specific_info": "specific_info"}], "action": {"modified_timestamp": "1641469450", "status": "pending"}, "attributes": {}}
cortx/ha/v1/cortx/ha/system/cluster/eabfcf05cf3f4c0db4fdde7ba7ae1285/site/1/rack/1/node/27d9b5122785444444444444444/health:{"events": [{"event_timestamp": "1641469283", "created_timestamp": "1641469283", "status": "failed", "specific_info": "specific_info"}, {"event_timestamp": "1641469138", "created_timestamp": "1641469139", "status": "online", "specific_info": "specific_info"}], "action": {"modified_timestamp": "1641469283", "status": "pending"}, "attributes": {}}
cortx/ha/v1/cortx/ha/system/cluster/eabfcf05cf3f4c0db4fdde7ba7ae1285/site/1/rack/1/node/6cbc441cad6b41e1a5d7bbd5fc3e5903/health:{"events": [{"event_timestamp": "1641468928", "created_timestamp": "1641468932", "status": "online", "specific_info": "specific_info"}], "action": {"modified_timestamp": "1641468932", "status": "pending"}, "attributes": {}}
cortx/ha/v1/cortx/ha/system/cluster/eabfcf05cf3f4c0db4fdde7ba7ae1285/site/1/rack/1/node/ead903f9bf9042e3b27c276291fff38c/health:{"events": [{"event_timestamp": "1641468927", "created_timestamp": "1641468928", "status": "online", "specific_info": "specific_info"}], "action": {"modified_timestamp": "1641468928", "status": "pending"}, "attributes": {}}
cortx/ha/v1/events/node/failed:["hare"]
cortx/ha/v1/events/node/online:["hare"]
cortx/ha/v1/events/subscribe/hare:["node/online", "node/failed"]
cortx/ha/v1/message_type/hare:ha_event_hare
    
  </code>
</pre>

### CODE REVIEW

File: ha/core/system_health/system_health_manager.py [hunk 1]
```diff
@@ -45,3 +45,15 @@ def set_key(self, key: str, value: str):
...
+        """
+        Check if key exists.
+        Args:
+            key (str): Key.
+        Return:
+            bool: True if key exists else False.
+        """
+        if self._store.key_exists(key):
+            return True
+        return False
```
Comment (mukhtar-inamdar) [inline ha/core/system_health/system_health_manager.py:59]: Add a new line at the end.

**Review by mukhtar-inamdar — APPROVED** [review]
add a new line at the end of the page: ha/core/system_health/system_health_manager.py.

**Review by Madhura-08 — APPROVED** [review]

File: ha/k8s_setup/ha_setup.py [hunk 2]
```diff
@@ -213,10 +213,6 @@ def process(self):
             data_pod_label = Conf.get(self._index, f'cortx{_DELIM}common{_DELIM}product_release')
             # TBD delete once data_pod_label is avilable from confstore
             data_pod_label = ['cortx-data', 'cortx-server']
-
-            # Time till when system health can be collected in bootstrap mode
-            timeout = Conf.get(self._index, f'cortx{_DELIM}common{_DELIM}product_release')
-            timeout = '10' # in seconds ;  temporary value till the same is avilabe in cluster.conf
             # Total number of pods for which health is to be maintained
             num_pods = Conf.get(self._index, f'cortx{_DELIM}common{_DELIM}product_release')
```
Comment (ArchanaLimaye) [inline ha/k8s_setup/ha_setup.py:217]: num_pods should also be removed

File: ha/k8s_setup/ha_setup.py [hunk 3]
```diff
@@ -235,7 +231,6 @@ def process(self):
                                               'consumer_id' : '2'},
                          'NODE': {'resource_type': 'node'},
                          'SYSTEM_HEALTH' : {'num_entity_health_events' : 2,
-                                            'sys_health_bootstrap_timeout' : timeout,
```
Comment (ArchanaLimaye) [inline ha/k8s_setup/ha_setup.py:238]: total_num_pods should also be removed

**Review by shaileshsvc — APPROVED** [review]
Approved

File: ha/k8s_setup/ha_setup.py (hunk 2 above)
Comment (akash2144) [reply ha/k8s_setup/ha_setup.py]: Removed.

File: ha/k8s_setup/ha_setup.py (hunk 3 above)
Comment (akash2144) [reply ha/k8s_setup/ha_setup.py:238]: Removed.

**Review by ArchanaLimaye — APPROVED** [review]

**Pull request merged**